src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -7,18 +7,115 @@
  */
 public class MyHashMap<K, V> implements MyMap<K, V> {
 
+    private static final int INITIAL_CAPACITY = 16;
+    private static final float LOAD_FACTOR = 0.75f;
+    private int size;
+    private int threshold;
+    private Entry<K, V>[] table;
+
+    public MyHashMap() {
+        table = new Entry[INITIAL_CAPACITY];
+        threshold = countThreshold(INITIAL_CAPACITY);
+    }
+
+    private static class Entry<K, V> {
+        private K key;
+        private V value;
+        private Entry<K, V> next;
+
+        public Entry(K key, V value) {
+            this.key = key;
+            this.value = value;
+        }
+    }
+
     @Override
     public void put(K key, V value) {
+        Entry<K, V> newEntry = new Entry<>(key, value);
+        if (key == null) {
+            putForNullKey(newEntry);
+            return;
+        }
+        int index = hash(key.hashCode());
+        if (table[index] == null) {
+            table[index] = newEntry;
+            checkSize(++size);
+            return;
+        }
+        Entry<K, V> entryByIndex = table[index];
+        while (entryByIndex != null) {
+            if (entryByIndex.key.equals(key)) {
+                entryByIndex.value = value;
+                return;
+            }
+            if (entryByIndex.next == null) {
+                entryByIndex.next = newEntry;
+                checkSize(++size);
+                return;
+            }
+            entryByIndex = entryByIndex.next;
+        }
+    }
+
+    private void putForNullKey(Entry<K, V> nullKeyEntry) {
+        if (table[0] == null) {
+            table[0] = nullKeyEntry;
+            size++;
+        } else {
+            table[0].value = nullKeyEntry.value;
+        }
+    }
+
+    private void checkSize(int size) {
+        if (size > threshold) {
+            resize();
+        }
+    }
+
+    private void resize() {
+        int newCapacity = table.length * 2;
+        Entry<K, V>[] oldTable = table;
+        table = new Entry[newCapacity];
+        threshold = countThreshold(newCapacity);
+        rehash(oldTable);
+    }
+
+    private void rehash(Entry<K, V>[] oldTable) {
+        size = 0;
+        for (Entry<K, V> entry : oldTable) {
+            while (entry != null) {
+                put(entry.key, entry.value);
+                entry = entry.next;
+            }
+        }
+    }
+
+    private int hash(int hashcode) {
+        return Math.abs(hashcode) % (table.length - 1) + 1;
```

Comment:
why do we add +1 here?

Reply:
We add +1 here because of table[0] is reserved for nullKey. And this hash() method should not return 0. For example if table.length is 16, this method should return values in range from 1 to 15;